plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/types/ClassPredicateEnhancer.kt

```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright 2010-2024 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.formver.embeddings.types
+
+import org.jetbrains.kotlin.formver.embeddings.expression.OperatorExpEmbeddings.StringLength
+
+/**
+ * Sometimes we want to enhance a predicate with additional assertions.
+ * Hence, we introduce `ClassPredicateEnhancer`.
+ */
+sealed class ClassPredicateEnhancer {
+    internal abstract fun applyAdditionalAssertions(builder: ClassPredicateBuilder)
+}
```

Comment:
Why not an interface?

Reply:
Oh, it's actually because internals are not allowed in interfaces. And it needs to be internal because builder was made internal. We actually could remove this restriction, I don't see a reason why it was internal in the first place.

Reply:
Sounds good to make it non-internal, I think we don't make that distinction anywhere else.